### PR TITLE
matching: make taskListID and taskListName partition aware

### PR DESCRIPTION
**Pull request by venkat1109**
This patch does 3 things:
 - makes taskListID type partition aware
-  introduces new type "qualifiedTaskListName" to represent internal task list names with metadata
- adds validation to prohibit tasklist names with reserved prefix

Related to #2098

### CODE REVIEW

File: service/history/decisionChecker.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ package history
 import (
 	"fmt"
```
Comment (shreyassrivatsan) [inline service/history/decisionChecker.go]: remove newline.. run `make fmt`

File: service/history/decisionChecker.go [hunk 2]
```diff
@@ -191,8 +195,8 @@ func (v *decisionAttrValidator) validateActivityScheduleAttributes(
 		return &workflow.BadRequestError{Message: "ScheduleActivityTaskDecisionAttributes is not set on decision."}
 	}
 
-	if attributes.TaskList == nil || attributes.TaskList.GetName() == "" {
-		return &workflow.BadRequestError{Message: "TaskList is not set on decision."}
+	if err := v.validateTaskListName(attributes.TaskList, ""); err != nil {
```
Comment (shreyassrivatsan) [inline service/history/decisionChecker.go]: can you set this to be a const `noTasklist` or something?

File: service/history/decisionChecker.go [hunk 3]
```diff
@@ -581,6 +581,31 @@ func (v *decisionAttrValidator) validateStartChildExecutionAttributes(
 	return nil
 }
 
+func (v *decisionAttrValidator) validateTaskListName(tl *workflow.TaskList, defaultVal string) error {
```
Comment (shreyassrivatsan) [inline service/history/decisionChecker.go]: This method does more than just `validate` - it changes the passed in task list with the `defaultVal`.. Perhaps `validateAndSetTaskListName` ?

File: service/matching/tasklist.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+// IsRoot returns true if this task list is a root partition
+func (tn *qualifiedTaskListName) IsRoot() bool {
+	return tn.partition == 0
+}
+
+// Parent returns the name of the parent task list
+// input:
+//   degree: Number of children at each level of the tree
+// Returns empty string if this task list is the root
+func (tn *qualifiedTaskListName) Parent(degree int) string {
```
Comment (shreyassrivatsan) [inline service/matching/tasklist.go]: `degree` must be > 1

Comment (shreyassrivatsan) [reply service/matching/tasklist.go]: Also, shouldn't degree be configured on the task list.. Why is this an input to the method?

File: service/history/decisionChecker.go (hunk 3 above)
Comment (venkat1109) [reply service/history/decisionChecker.go]: this code changed a bit, this comment is no longer relevant

File: service/matching/tasklist.go (hunk 4 above)
Comment (venkat1109) [reply service/matching/tasklist.go]: if its not an input, then this type needs to take in the dynamic config func as input in the constructor. the latter option results in passing dynamic config everywhere in the unit test. But the actual Parent() method is only going to be called in one or two places. So, I  decided to pass this a param to this function.

File: service/history/decisionChecker.go (hunk 1 above)
Comment (venkat1109) [reply service/history/decisionChecker.go]: weird, I did run make fmt

Comment (venkat1109) [reply service/history/decisionChecker.go]: fixed

File: service/history/decisionChecker.go (hunk 2 above)
Comment (venkat1109) [reply service/history/decisionChecker.go]: done

**Pull request merged**